pandas/_libs/lib.pyx

```diff
@@ -659,7 +660,11 @@ def array_equivalent_object(left: object[:], right: object[:]) -> bool:
                   and not (isinstance(x, type(y)) or isinstance(y, type(x)))):
                 # Check if non-scalars have the same type
                 return False
+            elif check_na_tuples_nonequal(x, y):
+                # We have tuples where one Side has a NA and the other side does not
+                return False
```

Comment:
comment that this is the only/main we we expect to get a TypeError

Reply:
Added